### PR TITLE
[Merged by Bors] - feat : Abelian groups has enough injectives

### PR DESCRIPTION
**Pull request by jjaassoonn**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Category/GroupCat/Injective.lean [hunk 1]
```diff
@@ -152,4 +157,275 @@ instance injective_of_divisible [DivisibleBy A ℤ] :
...
+    norm_cast at ineq0
+    rw [Int.abs_le_one_iff] at ineq0
+    pick_goal 2
+    · norm_num
+    pick_goal 2
+    · simp only [abs_pos, ne_eq, Int.cast_eq_zero]
+      rintro rfl
+      simp only [Int.cast_zero, add_zero, inv_eq_zero, Nat.cast_eq_zero] at H
+      linarith only [finite_order, H]
+    obtain (rfl|rfl|rfl) := ineq0
```
Comment (jjaassoonn) [inline Mathlib/Algebra/Category/GroupCat/Injective.lean:387]: This is silly

**Comment by jjaassoonn** [comment]
I looked back to one old Lean3 PR and adopted @alreadydone's idea

**Comment by jcommelin** [comment]
Please format your code according to the style guide. I see several pairs of `{}` that are not in the correct place.

**Comment by jjaassoonn** [comment]
> Please format your code according to the style guide. I see several pairs of `{}` that are not in the correct place.

These has been changed

File: Mathlib/Algebra/Category/GroupCat/Injective.lean [hunk 2]
```diff
@@ -152,4 +162,273 @@ instance injective_of_divisible [DivisibleBy A ℤ] :
             congr
 #align AddCommGroup.injective_of_divisible AddCommGroupCat.injective_of_divisible
 
+instance injective_ratCircle : CategoryTheory.Injective <| of <| ULift.{u} <| AddCircle (1 : ℚ) :=
+  have : Fact ((0 : ℚ) < 1) := ⟨by norm_num⟩
+  @injective_of_divisible _ _
+  { div := λ a n ↦ ULift.up (DivisibleBy.div a.down n)
+    div_zero := λ a ↦ by simp only [DivisibleBy.div_zero]; rfl
+    div_cancel := λ {n} a ha ↦ by ext1; exact DivisibleBy.div_cancel a.down ha }
```
Comment (eric-wieser) [inline Mathlib/Algebra/Category/GroupCat/Injective.lean:170]: This looks like a missing `Divisible` instance for `ULift`; can you add it in the appropriate place?

Comment (jjaassoonn) [reply Mathlib/Algebra/Category/GroupCat/Injective.lean]: I have moved this to `GroupTheory/Divisible.lean`

File: Mathlib/Algebra/Category/GroupCat/Injective.lean [hunk 3]
```diff
@@ -152,4 +162,269 @@ instance injective_of_divisible [DivisibleBy A ℤ] :
...
+    refine hf <| ULift.down_injective <| congr_fun h0
+      (@Injective.factorThru (ModuleCat ℤ) _ _ _ _ i2 f g hg).toAddMonoidHom
+
+section aux_defs
+
+variable {A_}
+variable (a : A_)
+
+/-- the `ℤ`-linear map of scalar multiplication (. • a)-/
+@[simps] def smulBya : ℤ →ₗ[ℤ] A_ where
```
Comment (jcommelin) [inline Mathlib/Algebra/Category/GroupCat/Injective.lean:212]: This is `LinearMap.toSpanSingleton ℤ (↑A_) a`

File: Mathlib/GroupTheory/Divisible.lean [hunk 4]
```diff
@@ -161,6 +162,16 @@ instance Prod.rootableBy : RootableBy (B × B') β where
 
 end Prod
 
+section ULift
+
+@[to_additive]
+instance ULift.rootableBy [RootableBy A α] : RootableBy (ULift A) α where
+  root := λ x a ↦ ULift.up <| RootableBy.root x.down a
+  root_zero := λ x ↦ ULift.ext _ _ <| RootableBy.root_zero x.down
+  root_cancel := λ _ h ↦ ULift.ext _ _ <| RootableBy.root_cancel _ h
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Divisible.lean:171]: The style guide says not to use `λ`, can you use `fun` instead?

File: Mathlib/Algebra/Category/GroupCat/Injective.lean [hunk 5]
```diff
@@ -152,4 +162,265 @@ instance injective_of_divisible [DivisibleBy A ℤ] :
...
+
+variable (A_ : AddCommGroupCat.{u})
+
+/-- the next term of `A`'s injective resolution is `∏_{A → ℚ/ℤ}, ℚ/ℤ`.-/
+def next : AddCommGroupCat.{u} := of <|
+  (A_ ⟶ of <| ULift <| AddCircle (1 : ℚ)) → AddCircle (1 : ℚ)
+
+instance : CategoryTheory.Injective <| next A_ :=
+  have : Fact ((0 : ℚ) < 1) := ⟨by norm_num⟩
+  @injective_of_divisible _ _ <| @Pi.divisibleBy _ _ _ _ _ _ <| λ f ↦ inferInstance
```
Comment (eric-wieser) [inline Mathlib/Algebra/Category/GroupCat/Injective.lean:179]: Are these underscores needed, or this this a lean3ism?

File: Mathlib/GroupTheory/Divisible.lean [hunk 6]
```diff
@@ -161,6 +162,16 @@ instance Prod.rootableBy : RootableBy (B × B') β where
 
 end Prod
 
+section ULift
+
+@[to_additive]
+instance ULift.rootableBy [RootableBy A α] : RootableBy (ULift A) α where
```
Comment (eric-wieser) [inline Mathlib/GroupTheory/Divisible.lean:168]:
```suggestion
instance ULift.instRootableBy [RootableBy A α] : RootableBy (ULift A) α where
```

File: Mathlib/GroupTheory/Divisible.lean (hunk 4 above)
Comment (eric-wieser) [reply Mathlib/GroupTheory/Divisible.lean:171]: (throughout)

File: Mathlib/Algebra/Category/GroupCat/Injective.lean (hunk 5 above)
Comment (jjaassoonn) [reply Mathlib/Algebra/Category/GroupCat/Injective.lean]: indeed they are not necessary

File: Mathlib/GroupTheory/Divisible.lean (hunk 4 above)
Comment (jjaassoonn) [reply Mathlib/GroupTheory/Divisible.lean:171]: done (I think)

File: Mathlib/GroupTheory/Divisible.lean [hunk 7]
```diff
@@ -161,6 +162,16 @@ instance Prod.rootableBy : RootableBy (B × B') β where
 
 end Prod
 
+section ULift
+
+@[to_additive]
+instance ULift.instRootableBy [RootableBy A α] : RootableBy (ULift A) α where
+  root := λ x a ↦ ULift.up <| RootableBy.root x.down a
+  root_zero := λ x ↦ ULift.ext _ _ <| RootableBy.root_zero x.down
+  root_cancel := λ _ h ↦ ULift.ext _ _ <| RootableBy.root_cancel _ h
```
Comment (alreadydone) [inline Mathlib/GroupTheory/Divisible.lean:171]: Still not changed ... and I think the following is more Lean 4 style
```suggestion
  root x a := ULift.up <| RootableBy.root x.down a
  root_zero x := ULift.ext _ _ <| RootableBy.root_zero x.down
  root_cancel _ h := ULift.ext _ _ <| RootableBy.root_cancel _ h
```

**Pull request closed without merging**